learning-sessions/Unit_testing_python_2026/example_code/tests/test_script.py

```
@@ -0,0 +1,44 @@
+        with pytest.raises(TypeError, match="Sequence must be a string"):
+            calculate_gc_content(12345)
+
+    def test_rounding(self):
+        """Test that the result is rounded correctly."""
+        # GC content of "GAT" is 1/3 = 33.3333...%
+        assert calculate_gc_content("GAT") == 33.33
+
+    def test_invalid_characters(self):
+        """Test that a ValueError is raised for sequences with invalid characters."""
+        with pytest.raises(ValueError, match="Sequence contains invalid characters. "
+                           "Only A, T, G, C are allowed."):
```

Ruff (0.15.2)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)